### ISSUE TITLE
bug: Projects without `environments` nor `default_environment` see error where `dev` environment is not found

### ISSUE THREAD

**Issue by edgarrmondragon** [meltano/meltano#6677]
### Meltano Version

2.5.0

### Python Version

NA

### Bug scope

Configuration (settings parsing, validation, etc.)

### Operating System

NA

### Description

Users with projects created from scratch (not using `meltano init`) or started before Environments where introduced may not be using `default_environment` nor `environments` in their `meltano.yml`.

This causes an error because in `2.5.0` we made `default_environment` a project setting that defaults to `dev`. Any project without a `dev` environment, as those exemplified above, will see an error because Meltano will be looking for `dev`.

### Code

```console
$ meltano install
Environment 'dev' was not found
```

This is reproducible with the following `meltano.yml`:

```yaml
plugins:
  loaders:
  - name: target-jsonl
    variant: andyh1203
    pip_url: target-jsonl
```

### Workaround

From AJ:

There should be two ways to work around this issue.

Running `meltano environment add dev` as a one-time fix will resolve the issue, even if no other changes are made and no environment-specific settings are added. 

Adding the `--no-environment` CLI arg will override the default `dev` environment name.

**Comment by pnadolny13** [comment]
@edgarrmondragon +1 I ran into this last night also.

**Comment by tayloramurphy** [comment]
@edgarrmondragon what's a reasonable resolution path here? Do we default to `default_environment: none` but put a prompt for users to consider adding environments? I don't think suddenly forcing everything through `dev` is a great experience on an existing project.

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon what's a reasonable resolution path here? Do we default to `default_environment: none` but put a prompt for users to consider adding environments? I don't think suddenly forcing everything through `dev` is a great experience on an existing project.

@tayloramurphy Yeah, I think defaulting to `none` in `settings.yml` is the right call. New projects created with `meltano init` will still get `default_environment: dev` and the corresponding environment.

**Comment by tayloramurphy** [comment]
@edgarrmondragon sounds good!

@aaronsteers @edgarrmondragon is this worth a 2.5.1? Feels like it since it's pretty big feeling bug.

**Comment by aaronsteers** [comment]
@tayloramurphy - I'm on the fence a bit due to the importance of environments for many of our default behaviors.

We had some discussion on this over here:

- #3200 

If there was a reasonable way to warn or even hard fail with a specific instruction, I would feel a bit better about that path forward.

There's another topic wrapped up here also, which is that some non-invocation commands shouldn't need an environment or apply the default one. 

I think these are the commands that probably shouldn't have a default environment applied:

- [add](https://docs.meltano.com/reference/command-line-interface#add)
- [config](https://docs.meltano.com/reference/command-line-interface#config)
- [discover](https://docs.meltano.com/reference/command-line-interface#discover)
- [environment](https://docs.meltano.com/reference/command-line-interface#environment)
- [init](https://docs.meltano.com/reference/command-line-interface#init)
- [install](https://docs.meltano.com/reference/command-line-interface#install)
- [invoke](https://docs.meltano.com/reference/command-line-interface#invoke)
- [lock](https://docs.meltano.com/reference/command-line-interface#lock)
- [remove](https://docs.meltano.com/reference/command-line-interface#remove)
- [job](https://docs.meltano.com/reference/command-line-interface#job)
- [schedule](https://docs.meltano.com/reference/command-line-interface#schedule)
- [select](https://docs.meltano.com/reference/command-line-interface#select)
- [state](https://docs.meltano.com/reference/command-line-interface#state)
- [test](https://docs.meltano.com/reference/command-line-interface#test)
- [ui](https://docs.meltano.com/reference/command-line-interface#ui)
- [user](https://docs.meltano.com/reference/command-line-interface#user)
- [upgrade](https://docs.meltano.com/reference/command-line-interface#upgrade)
- [version](https://docs.meltano.com/reference/command-line-interface#version)

This could be run without an environment but should apply a default when available:

- [elt](https://docs.meltano.com/reference/command-line-interface#elt)

This should really not be run without an environment, because behaviors like auto-state tracking won't work as expected when no environment is set:

- [run](https://docs.meltano.com/reference/command-line-interface#run)

**Comment by tayloramurphy** [comment]
@aaronsteers to be clear, I was proposing that if `default_environment` is not set in `meltano.yml` that we choose `none`, but I'm quite swayed by your argument. For commands that are "doing work" with a plugin, I'd argue we do want an environment always and should likely hard fail. 

My preference is that `elt`, `invoke`, `test`, and `run` are essentially treated the same from an environment perspective and everything else can be run w/o the default, though a user could set it if they wish. 

For the scope of this issue though, should we mainly release an path update around handling `default_environment` for those commands? Or something else?

**Comment by aaronsteers** [comment]
> For the scope of this issue though, should we mainly release an path update around handling `default_environment` for those commands? Or something else?

I think for the scope of this issue, I just would like to make sure we aren't onboarding a large set of users who will expect `meltano run` to work without any environments defined at all. We hadn't given it ton of attention as of yet, but with those design choices, we made it really _difficult_ (but not impossible) to invoke `meltano run` without a named environment. 

So, I do think we need to pair the proposed change with a stricter requirement in `meltano run` - and a simple guard clause and error message print would do the trick:

> The `meltano run` command requires a valid environment. You can create a dev environment by running `meltano environment create dev` and set it as default with `meltano config  default_environment dev`. For more information: https://docs.meltano.com/reference/command-line-interface#environment

If we pair these together, then I'm okay with the proposed change of default to `none`.

**Comment by tayloramurphy** [comment]
@aaronsteers I'm DM'ing you about this because I'm a bit confused on urgency. Will sync back our discussion.

**Comment by aaronsteers** [comment]
@tayloramurphy - I've added a "Workarounds" section above in the description. Since the workarounds seem pretty clear, I'm not seeing this as a high urgency item - or at least not high enough urgency that we'd want to introduce new tech debt or promote unsupported execution paths.

**Comment by pdebelak** [comment]
Jumping in here, I was surprised that this wasn't called out as a breaking change in the changelog. A `meltano.yml` file that used to work with `meltano install` gives this confusing error message when upgrading with no link to what an environment is or how to fix this error.

**Comment by aaronsteers** [comment]
First, I want to say thank you to everyone who's upvoted and commented here.

Second, given that this is affecting multiple users, I have bumped urgency to `High` and logged specific resolution steps:

- #6832 
- #6833 
- #6834 

And one more to update the changelog and release notes to document the breaking change per feedback above from @pdebelak :

- https://github.com/meltano/meltano/issues/6835

**Comment by alanmcruickshank** [comment]
Heya team - I've tried using the provided workaround from above and get the... same error. This is running on 2.7.2.

```
> pipenv run meltano environment add dev
Loading .env environment variables...
Need help fixing this problem? Visit http://melta.no/ for troubleshooting steps, or to
join our friendly Slack community.

Environment 'dev' was not found.
```

As for the adding `--no-environment` in all instances that we invoke meltano I'm not clear on how to do that in a production environment.

One thing it would definitely be useful to add is some mention of this issue in the upgrade guide. It's very confusing as a user to follow the migration guide and get an error for something which _isn't mentioned in the guide_.

I'll happily try a new release when the above PRs are released. Is there an ETA on release?

**Comment by alanmcruickshank** [comment]
Upgrade to 2.8.0 has helped with this issue! (caused another - but I think solvable) 👍

**Comment by aaronsteers** [comment]
@alanmcruickshank - Thanks very much for confirming. I'll close this issue as resolved as of `v2.8.0`. 👍
